cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java

```diff
@@ -112,9 +119,88 @@ private MetricsTable getOrCreateMetricsTable(String tableName, DatasetProperties
     return table;
   }
 
+  private MetricsTable getOrCreateResolutionMetricsTable(String v2TableName, String v3TableName,
+                                                 TableProperties.Builder props) {
+    try {
+      // metrics tables are in the system namespace
+      DatasetId v2TableId = NamespaceId.SYSTEM.dataset(v2TableName);
+
+      MetricsTable v2Table = null;
+      if (dsFramework.hasInstance(v2TableId)) {
+        v2Table = DatasetsUtil.getOrCreateDataset(dsFramework, v2TableId, MetricsTable.class.getName(),
+                                                  props.build(), null);
+      }
+
+      props.add(HBaseTableAdmin.PROPERTY_SPLITS,
+                GSON.toJson(getV3MetricsTableSplits(Constants.Metrics.METRICS_HBASE_SPLITS)));
+      props.add(Constants.Metrics.METICS_HBASE_TABLE_SPLITS, Constants.Metrics.METRICS_HBASE_SPLITS);
+      props.add(Constants.Metrics.METRICS_HBASE_MAX_SCAN_THREADS,
+                cConf.get(Constants.Metrics.METRICS_HBASE_MAX_SCAN_THREADS));
+
+      DatasetId v3TableId = NamespaceId.SYSTEM.dataset(v3TableName);
+      MetricsTable v3Table = DatasetsUtil.getOrCreateDataset(dsFramework, v3TableId, MetricsTable.class.getName(),
+                                                             props.build(), null);
+
+      if (v2Table != null) {
+        // the cluster is upgraded, so use Combined Metrics Table
+        return new CombinedHBaseMetricsTable(v2Table, v3Table);
+      }
+      return v3Table;
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private static byte[][] getV3MetricsTableSplits(int splits) {
+    RowKeyDistributorByHashPrefix rowKeyDistributor = new RowKeyDistributorByHashPrefix(
+      new RowKeyDistributorByHashPrefix.OneByteSimpleHash(splits));
+    return getSplitKeys(splits, splits, rowKeyDistributor);
+  }
+
+  private static byte[][] getSplitKeys(int splits, int buckets, AbstractRowKeyDistributor keyDistributor) {
```

Comment:
seems that this should be a method of the rowkeyDistributor. As it relies on knowledge about how salt keys are formed.